### ISSUE TITLE
UpdateRemotePlugins failing

### ISSUE THREAD

**Issue by timeyyy** [garyburd/lua-client#2]
i receive the following
```zsh
./plugin.lua:138: bad argument #1 to 'match' (string expected, got nil)
stack traceback:
	./neovim.lua:149: in function <./neovim.lua:147>
	[C]: in function 'match'
	./plugin.lua:138: in function 'get_plugin'
	./plugin.lua:164: in function <./plugin.lua:163>
	(tail call): ?
	[C]: in function 'xpcall'
	./neovim.lua:161: in function <./neovim.lua:160>
function remote#host#UpdateRemotePlugins[6]..<SNR>17_RegistrationCommands, line 20
Vim(let): Internal Error
remote/host: generated rplugin manifest :~/.local/share/nvim/rplugin.vim
```

**Comment by garyburd** [comment]
I cannot reproduce this.  Please give some details about your environment (OS, nvim version).

The error implies that nvim called the remote plugin `specs`  method without a path argument.

**Comment by timeyyy** [comment]
I am using ubuntu 16.04 with nvim 0.17 installed from source

**Comment by garyburd** [comment]
I cannot reproduce the issue with ubuntu 16.04 with nvim 0.17 installed from source.  

What does `echo globpath(&rtp, 'rplugin/lua/*.lua')` print?

**Comment by timeyyy** [comment]
Originally I got this errorr when I was using an older nvim version. I then used dein to install the plugin from my locally dir. In 0.17 it works if I install from github using dein it works. Could be a dein issue?

**Comment by timeyyy** [comment]
i am testing now with init.vim as 
```viml
if &compatible
    set nocompatible
endif
filetype off
set runtimepath^=$XDG_CONFIG_HOME/nvim/repos/github.com/Shougo/dein.vim

call dein#begin(expand('~/.cache/dein'))

call dein#add('garyburd/neols')
call dein#add('garyburd/lua-client')

call dein#end()
```

```bash
~/.cache/dein/.dein/rplugin/lua/langserver.lua
~/.cache/dein/.dein/rplugin/lua/langserver_spec.lua
~/.cache/dein/.dein/rplugin/lua/plugin.lua
```

**Comment by zchee** [comment]
@garyburd @timeyyy 
First of all, I haven't lua experience. It is the first time I write lua.

BTW, It seems to Lua 5.2 has been removed `setfenv()` and `getfenv()`(right?)
So, I was quick googling and found backward compatible local `setfenv` and `getfenv` function.
As an alternative,

`setfenv`
```lua
local function setfenv(fn, env)
  local i = 1
  while true do
    local name = debug.getupvalue(fn, i)
    if name == "_ENV" then
      debug.upvaluejoin(fn, i, (function()
        return env
      end), 1)
      break
    elseif not name then
      break
    end

    i = i + 1
  end

  return fn
end
```

`getfenv`
```lua
local function getfenv(fn)
  local i = 1
  while true do
    local name, val = debug.getupvalue(fn, i)
    if name == "_ENV" then
      return val
    elseif not name then
      break
    end
    i = i + 1
  end
end
```

references from http://leafo.net/guides/setfenv-in-lua52-and-above.html

After that, adding `setfenv` function to `lua-client/plugin.lua`,

```diff
diff --git a/plugin.lua b/plugin.lua
index 925d893..fc4a6a3 100644
--- a/plugin.lua
+++ b/plugin.lua
@@ -71,6 +71,24 @@ local function define(specs, handlers, stype, opts)
   }
 end
 
+local function setfenv(fn, env)
+  local i = 1
+  while true do
+    local name = debug.getupvalue(fn, i)
+    if name == "_ENV" then
+      debug.upvaluejoin(fn, i, (function()
+        return env
+      end), 1)
+      break
+    elseif not name then
+      break
+    end
+
+    i = i + 1
+  end
+
+  return fn
+end
 
 local function new_plugin(dir, nvim)
   local plugin = setmetatable({
```

It seems to works fine only `:UpdateRemotePlugins` command. I haven't test another things(such as whether the works fine neols.)

I hope you find it informative.

**Comment by justinmk** [comment]
I'm not sure that it makes sense to support lua 5.2+, at least not initially. In the meantime I would say if lua 5.2+ is detected it should fail early with an informative error.

**Comment by garyburd** [comment]
LuaJIT is required at this time.  I added issue #3 for the informative error. The runtime environment is still TBD, but I think we do need to target a specific version of Lua.

@timeyyy Lua 5.2 replaced `setenv` with the `env` argument to `load` and friends.  The 5.2 compatible code is:

     assert(pcall(assert(loadfile(path, nil, env))))

**Comment by garyburd** [comment]
@timeyyy I cannot reproduce the issue with the posted init.vim.  

I cannot reproduce the output from `echo globpath(&rtp, 'rplugin/lua/*.lua')` with the posted init.vim.  The unexpanded `~` in the filenames will be the next issue after we figure out why the `specs` API is called with no arguments or a nil argument.

**Comment by timeyyy** [comment]
The ~ was just a short hand for /home/timeyyy,

**Comment by garyburd** [comment]
What does `lua -v` print?

**Comment by timeyyy** [comment]
`lua -v` -> `Lua 5.1.5  Copyright (C) 1994-2012 Lua.org, PUC-Rio`
`luajit -v` -> `LuaJIT 2.0.4 -- Copyright (C) 2005-2015 Mike Pall. http://luajit.org/`

**Comment by timeyyy** [comment]
Lua

**Comment by timeyyy** [comment]
I feel like this is something to do with how i setup my environment, Are you also using an Ubuntu environment? Would you be able to give a step by step on how you install luajit and luarocks?

**Comment by garyburd** [comment]
The plugin assumes that LuaJIT is invoked with the `lua` command (see source [here](https://github.com/garyburd/lua-client/blob/713b19d80851d253b760d23640e24a92ce3b7bcd/plugin/luahost.vim#L29)).  

I abandoned this plugin because installing the dependencies are a significant barrier to it's use. I suggest following tjdevries's lang server work.

**Comment by justinmk** [comment]
@garyburd Which dependencies? Nvim can be used as the lua runtime, and we can ship necessary lua modules with Nvim in `runtime/`.

**Comment by garyburd** [comment]
@justinmk Let me rephrase. I set the project aside because of the dependency issue and have not had time to come back to it since Lua was added to Nvim. I will probably not have time for the project until October.

If somebody has interest in this project, then they should feel free to clone and run with it.

**Comment by garyburd** [comment]
Fixed by 79a6a632f7d8d4dcf9ad5c1f62e763d91cbf3f59.
